rucio-dataset-mon/src/go/rucio-go-datatable/controllers/dataset_controller.go

```diff
@@ -0,0 +1,172 @@
+
+func CORS() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		//c.Writer.Header().Set("Content-Type", "application/json")
+		c.Writer.Header().Set("Access-Control-Allow-Origin", "*")
+		c.Writer.Header().Set("Access-Control-Allow-Credentials", "true")
+		c.Writer.Header().Set("Access-Control-Allow-Headers", "Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token, Authorization, accept, origin, Cache-Control, X-Requested-With")
+		c.Writer.Header().Set("Access-Control-Allow-Methods", "POST, OPTIONS, GET, PUT, DELETE")
+		if c.Request.Method == "OPTIONS" {
+			c.AbortWithStatus(204)
+			return
+		}
+		c.Next()
+	}
+}
+
+//func StrToInt64(s string, defaultVal int64) int64 {
+//	// DataTable draw parameter
+//	intVal, err := strconv.ParseInt(s, 10, 64)
+//	if err != nil {
+//		log.Printf("String conversion error. value: %s, error: %s", s, err)
+//		return defaultVal
+//	}
+//	return intVal
+//}
+
+//func QueryOrder(c *gin.Context) string {
+//	columnMap := map[string]string{
+//		"0": "id",
+//		"1": "dataset",
+//		"2": "rse",
+//		"3": "size",
+//	}
+//
+//	column := c.DefaultQuery("order[0][column]", "0")
+//	dir := c.DefaultQuery("order[0][dir]", "desc")
+//	orderString := columnMap[column] + " " + dir
+//
+//	return orderString
+//}
+
+func getTotalRecCount(ctx context.Context) int64 {
+	countTotal, err := datasetCollection.CountDocuments(ctx, bson.M{})
+	if err != nil {
+		panic(err)
+	}
+	return countTotal
+}
+
+func convertOrderStrToMongoInt(dir string) int {
+	switch strings.ToLower(dir) {
+	case "asc":
+		return 1
+	case "desc":
+		return -1
+	default:
+		return 0
+	}
+}
+
+// generateSortOrder creates sort oder Mongo query by iterating over DT response
+func generateSortOrder(dtReceivedJson models.DtReceivedJson) *options.FindOptions {
+	orders := dtReceivedJson.Orders
+	columns := dtReceivedJson.Columms
+	findOpts := options.Find()
+
+	for _, order := range orders {
+		intSortDir := convertOrderStrToMongoInt(order.Dir)
+		if intSortDir != 0 {
+			findOpts = findOpts.SetSort(
+				bson.D{{
+					columns[order.Column].Data, // column name
+					intSortDir,                 // column direction as int value
+				}},
+			)
+		}
+	}
+	return findOpts
+}
+
+func generateFindQuery(dtReceivedJson models.DtReceivedJson) []bson.M {
+	generalSearch := dtReceivedJson.Search
+	var findQuery []bson.M
+	findQuery = append(findQuery, bson.M{"dataset": primitive.Regex{Pattern: generalSearch.Value, Options: "im"}})
+	// i: case insentive, m: can use ^ and $. Ref: // https://www.mongodb.com/docs/v5.0/reference/operator/query/regex/
+	// TODO add individual column search
+	return findQuery
+}
+
+func GetAllDatasets() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		var datasets []models.Dataset
+		defer cancel()
+
+		// === ~~~~~~ Decode incoming DataTable json ~~~~~~ ===
+		body, err := ioutil.ReadAll(c.Request.Body)
```

Comment:
in go 1.18 most of ioutil functions relocated to os module, so you should use `os.ReadeAll`, etc.